File: src/components/dapps/signDataWarningPolicy.test.ts
import type { UnifiedSignDataPayload } from '../../api/dappProtocols/types';

import { getSignDataWarningKinds } from './signDataWarningPolicy';

const PAYLOADS = {
  text: { type: 'text', text: 'Message' },
  binary: { type: 'binary', bytes: '00' },
  cell: { type: 'cell', schema: 'root$_ = Root;', cell: 'te6ccgEBAQEAAgAAAA==' },
  eip712: {
    type: 'eip712',
    domain: {},
    types: {},
    primaryType: 'Message',
    message: {},
  },
} satisfies Record<string, UnifiedSignDataPayload>;

describe('signData warning policy', () => {
  it.each([
    ['text', PAYLOADS.text],
    ['binary', PAYLOADS.binary],
    ['EIP-712', PAYLOADS.eip712],
    ['parsed or fallback TL-B cell', PAYLOADS.cell],
  ])('shows exactly one generic trust warning for %s', (_name, payload) => {
    expect(getSignDataWarningKinds(payload)).toEqual(['genericTrust']);
  });

  it('does not render a warning before a signData payload is available', () => {
    expect(getSignDataWarningKinds(undefined)).toEqual([]);
  });
});

File: mobile/android/air/SubModules/UITonConnect/src/test/java/org/mytonwallet/app_air/uitonconnect/viewControllers/SignDataWarningPolicyTest.kt
package org.mytonwallet.app_air.uitonconnect.viewControllers

import org.junit.Assert.assertEquals
import org.junit.Test
import org.mytonwallet.app_air.walletcore.moshi.MSignDataPayload

class SignDataWarningPolicyTest {
    @Test
    fun `every sign data payload shows exactly one generic trust warning`() {
        val payloads = listOf(
            MSignDataPayload.SignDataPayloadText("Message"),
            MSignDataPayload.SignDataPayloadBinary("00"),
            MSignDataPayload.SignDataPayloadEip712(
                domain = emptyMap(),
                types = emptyMap(),
                primaryType = "Message",
                message = emptyMap()
            ),
            cellPayload
        )

        payloads.forEach { payload ->
            assertEquals(
                listOf(SignDataWarningKind.GENERIC_TRUST),
                signDataWarningKinds(payload)
            )
        }
    }

    private val cellPayload = MSignDataPayload.SignDataPayloadCell(
        schema = "root\$_ = Root;",
        cell = "te6ccgEBAQEAAgAAAA=="
    )
}
